紀錄/32 20211102 binary tree.md

```diff
@@ -26,3 +26,6 @@ https://en.wikipedia.org/wiki/Binary_tree
 * 	MEDIUM	117	Populating Next Right Pointers in Each Node II	https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii
 * 	MEDIUM	236	Lowest Common Ancestor of a Binary Tree	https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree
 * 	HARD	297	Serialize and Deserialize Binary Tree	https://leetcode.com/problems/serialize-and-deserialize-binary-tree
+
+--- 
+Notes:
```

Comment:
```suggestion
Notes:

```typescript=
/** Iteration in TS */
function preorderTraversal (root: TreeNode | null): number[] {
    let results: number[] = [];
    let rightNodeStack: TreeNode[] = [null];
    let curr = root;
    while (curr != null) {
        results.push(curr.val);
        if (curr.right != null) rightNodeStack.push(curr.right);
        if (curr.left != null) curr = curr.left;
        else curr = rightNodeStack.pop();
    }
    
    return results;
}

/** Recursion in TS */
function preorderTraversal (root: TreeNode | null): number[] {
    let results: number[] = [];
    recursiveHelper(root);
    return results;
   
    function recursiveHelper (node: TreeNode){
        if (node == null) return;
        results.push(node.val);
        recursiveHelper(node.left);
        recursiveHelper(node.right);
    }
};

```


```